### ISSUE TITLE
Enhancement to querying support 2 - Filtering Operators

### ISSUE THREAD

**Issue by abhishek-ram** [proteanhq/protean#68]
We currently do not support any operators other than the `=` filter. This is very limiting and requires us to use third party code of underlying adapter implementations like `sqlalchemy`. This ticket proposes to introduce operators that will then need to be converted to adapter specific code in the adapter packages. As an example we can support the following queries

```
Entry.filter(created_at__gte=yesterday)
Account.filter(name__startswith='John')
```

We can take a look at libraries Django and Pewee to come up with our implementation.

**Comment by abhishek-ram** [comment]
@subhashb One way we could do this is that the `QuerySet` class will convert all keyword arguments for filter and exclude to a 3 elements tuple of `(field, operator, value)` where the operator will be one from a list of constants. So the filter `age__gte=24` will be converted to `(age, op.GTE, 24)`.

The adapter repositories will the need to convert them to adapter specific filters at the time of evaluation.

**Comment by subhashb** [comment]
@abhishek-ram Here's my proposal:

Supported filtering/query operators are:
* `exact` - Default operator, equivalent to `eq` for all objects
* `iexact` - case-insensitive match, if String. Will throw error for non-character types
* `contains`
* `icontains`
* `in` - Applicable to iterable types
* `gte`
* `gt`
* `lte`
* `lt`
* *NOT* is supported through `exclude()`

More operators to come in later.

Conjunctions:
* All attributes in `.filter()` are **AND**ed
* Multiple filter clauses get collapsed together as part of a single AND clause
* **OR** is possible with an explicit call to `.or_` method
* _Example 1_: `Account.query.filter(first_name__iexact='Subhash', last_name__iexact=='Bhushan').or(full_name__iexact='Subhash Bhushan)`
* _Example 2_: `Account.query.filter(first_name__iexact='Subhash').filter(last_name__iexact=='Bhushan').or(full_name__iexact='Subhash Bhushan)`
* _Example 3_: `Account.query.filter(first_name__iexact='Subhash', last_name__iexact=='Bhushan').or(full_name__iexact='Subhash Bhushan).order_by('email')`

Joins:
* Will be supported as part of object attributes, like `Account.addresses` or `Post.filter(id=22334).account`
* Child/Parent object queries are supported via connected attribute names like `Account.query.filter(post.id==3345)`

Explcity Query:
* An explicit `where()` method will support direct query embedding, like `Account.where('first_name LIKE %s' % fn)`

I am sure we can enhance to support more complex queries, but this basic set should allow us to define most business rules. Do you see anything missing right off the bat?

Let's discuss.

**Comment by subhashb** [comment]
Closed with #72
